pkg/kernelcache/kext.go

```diff
@@ -2,69 +2,220 @@ package kernelcache
+	kc, err := NewKernelCache(kernel)
+	if err != nil {
+		return err
+	}
+	defer kc.Close()
+
+	prelink, err := kc.PrelinkInfoDict()
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("FOUND:", len(prelink.PrelinkInfoDictionary))
+	prelink.ForeachBundle(func(bundle * CFBundle) error {
+		fmt.Printf("%s (%s)\n", bundle.ID, bundle.Version)
+		return nil
+	})
+
+	return nil
 }
 
-type CFBundle struct {
-	Name                  string `plist:"CFBundleName,omitempty"`
-	ID                    string `plist:"CFBundleIdentifier,omitempty"`
-	InfoDictionaryVersion string `plist:"CFBundleInfoDictionaryVersion,omitempty"`
-	CompatibleVersion     string `plist:"OSBundleCompatibleVersion,omitempty"`
-	Version               string `plist:"CFBundleVersion,omitempty"`
-	Required              string `plist:"OSBundleRequired,omitempty"`
-	Executable            string `plist:"CFBundleExecutable,omitempty"`
-	OSKernelResource      bool   `plist:"OSKernelResource,omitempty"`
-	GetInfoString         string `plist:"CFBundleGetInfoString,omitempty"`
-	AllowUserLoad         bool   `plist:"OSBundleAllowUserLoad,omitempty"`
-	Signature             string `plist:"CFBundleSignature,omitempty"`
-	PackageType           string `plist:"CFBundlePackageType,omitempty"`
-	DevelopmentRegion     string `plist:"CFBundleDevelopmentRegion,omitempty"`
-	ShortVersionString    string `plist:"CFBundleShortVersionString,omitempty"`
+// unSeek performs thingy() and seeks fh back to its position before thingy() was performed.
+func unSeek(fh * os.File, thingy func() error) error {
+	var startSeek int64
+	var err error
+	if startSeek, err = fh.Seek(0, io.SeekCurrent); err != nil {
+		return err
+	}
+
+	err = thingy()
+
+	if _, seekBackErr := fh.Seek(startSeek, io.SeekStart); seekBackErr != nil {
+		return seekBackErr
+	}
+
+	return err
 }
 
-// KextList lists all the kernel extensions in the kernelcache
-func KextList(kernel string) error {
-	m, err := macho.Open(kernel)
+type Segment64 struct {
+	LoadCmd	  uint32
+	Len       uint32
+	Name      [16]byte
+	Addr      uint64
+	Memsz     uint64
+	Offset    uint64
+	Filesz    uint64
+	Maxprot   uint32
+	Prot      uint32
+	Nsect     uint32
+	Flag      uint32
+}
+
+type Section64 struct {
+	Name      [16]byte
+	Seg       [16]byte
+	Addr      uint64
+	Size      uint64
+	Offset    uint32
+	Align     uint32
+	Reloff    uint32
+	Nreloc    uint32
+	Flags     uint32
+	Reserved1 uint32
+	Reserved2 uint32
+	Reserved3 uint32
+}
+
+func roundUp64(x, y uint64) uint64 {
+	return (x + y) & (^(y - 1))
+}
+
+// ExtractKext locates the specified kext and writes it to a file with the same name.
+func ExtractKext(kernel, kextName string) error {
+	fh, err := os.Create(kextName)
 	if err != nil {
 		return err
 	}
-	defer m.Close()
 
-	for _, sec := range m.Sections {
-		if sec.Seg == "__PRELINK_INFO" && sec.Name == "__info" {
-			f, err := os.Open(kernel)
-			if err != nil {
-				return err
+	defer fh.Close()
+
+	kc, err := NewKernelCache(kernel)
```

Comment:
Check failure on line 92 in pkg/kernelcache/kext.go

 
GitHub Actions
/ build (windows-latest)

pkg/kernelcache/kext.go#L92
undefined: NewKernelCache